pkg/jobWorker/job.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	if err = ns.AddResourceControl(cgroupName, ns.CPU_WEIGHT_File, strconv.Itoa(int(job.config.CPU*100))); err != nil {
+		log.Printf("could not add resources into controller:%s, %v", ns.CPU_WEIGHT_File, err)
+		return fmt.Errorf("error starting command: %w", err)
+	}
+	if err = ns.AddResourceControl(cgroupName, ns.MEMORY_HIGH_File, strconv.FormatInt(job.config.MemBytes, 10)); err != nil {
+		return fmt.Errorf("could not add resources into controller:%s, %v", ns.MEMORY_HIGH_File, err)
+	}
+	//if err = ns.AddResourceControl(cgroupName, ns.IO_WEIGHT_File, strconv.FormatInt(job.config.IOBytesPerSecond, 10)); err != nil {
+	//	return fmt.Errorf("could not add resources into controller:%s, %v", ns.IO_WEIGHT_File, err)
+	//}
+
+	//provide the file descriptor to cmd.Run so that it can add the new PID to the control group
+	if err = ns.AddProcess(cgroupName, cmd); err != nil {
+		log.Printf("Error AddProcess /proc - %s\n", err)
+		os.Exit(1)
+	}
+
+	if err = ns.MountProc(); err != nil {
+		log.Printf("Error mounting /proc - %s\n", err)
+		os.Exit(1)
+	}
+
+	log.Printf("starting job:%s, cmd:%s", job, cmd.String())
+	if err = cmd.Start(); err != nil {
+		return fmt.Errorf("error starting command: %w", err)
+	}
+	job.cmd = cmd
+	job.status.State = JOB_STATUS_RUNNING
+
+	// run the command in a Goroutine so that Start can return immediately
+	go func() {
+		// Use cmd.Process.Wait() instead of cmd.Wait() since cmd.Wait() is not thread safe
+		// and we do not want to hold the mutex while waiting for the process to exit.
+		// So instead we use cmd.Process.Wait() and store the result in j.processState to mimic what
+		// cmd.Wait would do.
+		// This prevents concurrency issues when a user calls Start(), the command quickly exits (updating the
+		// process state), and the user invokes Status().
+		processState, err := job.cmd.Process.Wait()
+
+		job.mutex.Lock()
+		defer job.mutex.Unlock()
+
+		job.processState = processState
+		job.status.ExitCode = job.processState.ExitCode()
+		if job.status.State == JOB_STATUS_RUNNING {
+			job.status.State = JOB_STATUS_COMPLETED
+		}
+
+		if err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error running command: %s\n", err)
+		}
+
+		if err == nil && !job.processState.Success() {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error running command: %s\n", &exec.ExitError{ProcessState: job.processState})
+		}
+
+		// close the output, so that any readers of the output know the process has exited and will no longer
+		// block waiting for new output
+		if err = job.output.Close(); err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error closing output: %s\n", err)
+		}
+
+		// do not close the cgroup.procs file until after the process has exited
+		if err = ns.DeleteCGroup(cgroupName); err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error closing cgroup: %s\n", err)
+		}
```

Comment:
This error hanging is quite confusing Does internal server error should be propagate to the user  and mixed with job error ?

Reply:
@smallinsky Ok. changed in more appropriate error handling and propagate error to the library consumer.
<img width="1045" alt="Screenshot 2024-09-04 at 6 16 14 PM" src="https://github.com/user-attachments/assets/1f0c042e-88e6-4a36-9c11-19384ef556cb">